File: sdk/python/hitechcloud/exceptions.py
"""HiTechCloud SDK Exceptions"""


class HiTechCloudError(Exception):
    """Base exception for HiTechCloud SDK"""

    def __init__(self, message: str, status_code: int = None, errors: list = None):
        self.message = message
        self.status_code = status_code
        self.errors = errors or []
        super().__init__(self.message)


class AuthenticationError(HiTechCloudError):
    """Raised when authentication fails (401)"""
    pass


class AuthorizationError(HiTechCloudError):
    """Raised when access is forbidden (403)"""
    pass


class NotFoundError(HiTechCloudError):
    """Raised when resource is not found (404)"""
    pass


class RateLimitError(HiTechCloudError):
    """Raised when rate limit is exceeded (429)"""

    def __init__(self, message: str, retry_after: int = None, **kwargs):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after


class ValidationError(HiTechCloudError):
    """Raised when request validation fails (422)"""
    pass


class ServerError(HiTechCloudError):
    """Raised when server returns 5xx error"""
    pass


class NotImplementedError(HiTechCloudError):
    """Raised when endpoint is not implemented in OTE (501)"""
    pass


ERROR_MAP = {
    401: AuthenticationError,
    403: AuthorizationError,
    404: NotFoundError,
    422: ValidationError,
    429: RateLimitError,
    500: ServerError,
    501: NotImplementedError,
    503: ServerError,
}

File: sdk/ruby/lib/hitechcloud/exceptions.rb
# frozen_string_literal: true

module HiTechCloud
  class HiTechCloudError < StandardError
    attr_reader :status_code, :data

    def initialize(message, status_code: nil, data: nil)
      super(message)
      @status_code = status_code
      @data = data
    end
  end

  Error = HiTechCloudError

  class AuthenticationError < HiTechCloudError; end
  class AuthorizationError < HiTechCloudError; end
  class NotFoundError < HiTechCloudError; end

  class RateLimitError < HiTechCloudError
    attr_reader :retry_after

    def initialize(message, retry_after: 60, **opts)
      super(message, **opts)
      @retry_after = retry_after
    end
  end

  class ValidationError < HiTechCloudError; end
  class ServerError < HiTechCloudError; end

  def self.map_error(status_code, data)
    message = data['error'] || data['message'] || 'API error'

    case status_code
    when 401
      raise AuthenticationError.new(message, status_code: status_code, data: data)
    when 403
      raise AuthorizationError.new(message, status_code: status_code, data: data)
    when 404
      raise NotFoundError.new(message, status_code: status_code, data: data)
    when 422
      raise ValidationError.new(message, status_code: status_code, data: data)
    when 429
      raise RateLimitError.new(message, status_code: status_code, data: data)
    else
      if status_code >= 500
        raise ServerError.new(message, status_code: status_code, data: data)
      end
      raise HiTechCloudError.new(message, status_code: status_code, data: data)
    end
  end
end
